pyamg/amg_core/ruge_stuben.h

```diff
@@ -956,345 +943,1339 @@ void cr_helper(const I A_rowptr[],  const int A_rowptr_size,
+                                      const T Sx[], const int Sx_size,
+                                      const I splitting[], const int splitting_size,
+                                      const I Pp[], const int Pp_size,
+                                            I Pj[], const int Pj_size,
+                                            T Px[], const int Px_size)
+{
+    for (I i = 0; i < n_nodes; i++) {
+        // If node i is a C-point, then set interpolation as injection
+        if(splitting[i] == C_NODE) {
+            Pj[Pp[i]] = i;
+            Px[Pp[i]] = 1;
+        } 
+        // Otherwise, use RS standard interpolation formula
+        else {
+
+            // Calculate denominator
+            T denominator = 0;
+
+            // Start by summing entire row of A
+            for (I mm = Ap[i]; mm < Ap[i+1]; mm++) {
+                denominator += Ax[mm];
+            }
+
+            // Then subtract off the strong connections so that you are left with 
+            // denominator = a_ii + sum_{m in weak connections} a_im
+            for (I mm = Sp[i]; mm < Sp[i+1]; mm++) {
+                if ( Sj[mm] != i ) {
+                    denominator -= Sx[mm]; // making sure to leave the diagonal entry in there
+                }
+            }
+
+            // Set entries in P (interpolation weights w_ij from strongly connected C-points)
+            I nnz = Pp[i];
+            for (I jj = Sp[i]; jj < Sp[i+1]; jj++) {
+
+                if (splitting[Sj[jj]] == C_NODE) {
+
+                    // Set temporary value for Pj as global index, j. Will be mapped to
+                    // appropriate coarse-grid column index after all data is filled in. 
+                    Pj[nnz] = Sj[jj];
+                    I j = Sj[jj];
+
+                    // Initialize numerator as a_ij
+                    T numerator = Sx[jj];
+
+                    // Sum over strongly connected fine points
+                    for (I kk = Sp[i]; kk < Sp[i+1]; kk++) {
+                        if ( (splitting[Sj[kk]] == F_NODE) && (Sj[kk] != i) ) {
+                            
+                            // Get column k and value a_ik
+                            I k = Sj[kk];
+                            T a_ik = Sx[kk];
+
+                            // Get a_kj (have to search over k'th row in A for connection a_kj)
+                            T a_kj = 0;
+                            T a_kk = 0;
+                            for (I search_ind = Ap[k]; search_ind < Ap[k+1]; search_ind++) {
+                                if (Aj[search_ind] == j) {
+                                    a_kj = Ax[search_ind];
+                                }
+                                else if (Aj[search_ind] == k) {
+                                    a_kk = Ax[search_ind];
+                                }
+                            }
+
+                            // If sign of a_kj matches sign of a_kk, ignore a_kj in sum
```

Comment:
Is this `signof(a_kj)` block, the only difference with `rs_standard_interpolation_pass2` ?  
If so, could these functions be combined, but have a boolean "modified" flag passed in? 

So, we'd have
`if (modified && (signof(a_kj) == signof(a_kk)) { `
`   a_kj = 0;`
`}`

Would this work?